File: Design_Patterns/SolidPrinciples/Dependency_Inversion_Principle/05_DIP.js
// TITLE: Dependency Inversion Principle (DIP) - MacBook Example

// ========================================
// THE PROBLEM (VIOLATION):
// MacBook depends on concrete WiredKeyboard
// ========================================

class WiredKeyboardBad {
    type() {
        console.log("Typing on wired keyboard");
    }
}

class MacBookBad {
    constructor() {
        // Hard dependency!
        this.keyboard = new WiredKeyboardBad();
    }

    useKeyboard() {
        this.keyboard.type();
    }
}

// ========================================
// THE SOLUTION:
// Depend on abstraction (via dependency injection)
// ========================================

class WiredKeyboard {
    type() {
        console.log("Typing on wired keyboard");
    }
}

class BluetoothKeyboard {
    type() {
        console.log("Typing on Bluetooth keyboard");
    }
}

class MacBook {
    constructor(keyboard) { // Dependency injection!
        this.keyboard = keyboard;
    }

    useKeyboard() {
        this.keyboard.type();
    }
}

// Usage
console.log("=== PROBLEM (DIP Violation) ===");
const macbookBad = new MacBookBad();
macbookBad.useKeyboard(); // Stuck with wired keyboard!

console.log("\n=== SOLUTION (DIP Compliant) ===");
const wired = new WiredKeyboard();
const macbook1 = new MacBook(wired);
macbook1.useKeyboard();

const bluetooth = new BluetoothKeyboard();
const macbook2 = new MacBook(bluetooth);
macbook2.useKeyboard();

File: Design_Patterns/SolidPrinciples/Dependency_Inversion_Principle/05_DIP.py
from abc import ABC, abstractmethod

# TITLE: Dependency Inversion Principle (DIP) - MacBook Example

# ========================================
# THE PROBLEM (VIOLATION):
# MacBook depends on concrete WiredKeyboard
# ========================================

class WiredKeyboardBad:
    def type(self):
        print("Typing on wired keyboard")

class MacBookBad:
    def __init__(self):
        # Hard dependency!
        self.keyboard = WiredKeyboardBad()
    
    def use_keyboard(self):
        self.keyboard.type()

# ========================================
# THE SOLUTION:
# Depend on abstraction
# ========================================

class Keyboard(ABC):
    @abstractmethod
    def type(self): pass

class WiredKeyboard(Keyboard):
    def type(self):
        print("Typing on wired keyboard")

class BluetoothKeyboard(Keyboard):
    def type(self):
        print("Typing on Bluetooth keyboard")

class MacBook:
    def __init__(self, keyboard: Keyboard): # Dependency injection!
        self.keyboard = keyboard
    
    def use_keyboard(self):
        self.keyboard.type()

if __name__ == "__main__":
    print("=== PROBLEM (DIP Violation) ===")
    macbook_bad = MacBookBad()
    macbook_bad.use_keyboard() # Stuck with wired keyboard!
    
    print("\n=== SOLUTION (DIP Compliant) ===")
    wired = WiredKeyboard()
    macbook1 = MacBook(wired)
    macbook1.use_keyboard()
    
    bluetooth = BluetoothKeyboard()
    macbook2 = MacBook(bluetooth)
    macbook2.use_keyboard()
